### ISSUE TITLE
Safe shut down for sequence of requests

### ISSUE THREAD

**Issue by ctrlok** [plabayo/tokio-graceful#7]
Hello. First of all — thanks for the great job. 

I'm trying to implement a graceful shutdown for my app. 

The idea of the app is pretty simple: 
1. It gets a batch of messages from the queue 
2. it processes each message one by one 

The first part of processing may take quite some time and can be interrupted by a shutdown signal.
The second part of processing is fast and better not to be interrupted (store result to db/queue) 
If we shut down the app - each unprocessed message should placed back in the queue (i.e. canceling job)

For example, let's say we have [0..10] as the naive implementation of batch messages. The code will look like this: 

```rust
use std::time::Duration;
use tokio::task;
use tokio::time::sleep;
use tokio_graceful::Shutdown;

#[tokio::main]
async fn main() {
    let shutdown = Shutdown::default();
    let guard1 = shutdown.guard();
    task::spawn(async move {
        for i in 0..10 {
            tokio::select! {
                _ = guard1.cancelled() => {
                    canceling(i).await;
                },
                result = non_mandatory_work(i) => {
                    mandatory_work(result).await;
                },
            }
        }
    });

    // Wait for shutdown signal
    shutdown.shutdown().await;
}

async fn canceling(i: u32) {
    println!("Canceling after {}", i);
}

async fn non_mandatory_work(i: u32) -> u32 {
    println!("Non-mandatory work started for {}", i);
    sleep(Duration::from_secs(2)).await;
    println!("Non-mandatory work done for {}", i);
    i
}

async fn mandatory_work(i: u32) {
    println!("Mandatory work started for {}", i);
    sleep(Duration::from_secs(2)).await;
    println!("Mandatory work done for {}", i);
}
```

In this code, tokio cancels `non_mandatory_work` after shutdown. And `mandatory_work` never happens because `guard1.cancelled() ` is faster then `non_mandatory_work` 

<details>

<summary>Still, it starts `non_mandatory_work` as well and just drops it before it can finish. 
See output (^C is cancel) </summary>

```
Non-mandatory work started for 0
^C Canceling after 0
Non-mandatory work started for 1
Canceling after 1
Non-mandatory work started for 2
Canceling after 2
Non-mandatory work started for 3
Canceling after 3
Canceling after 4
Non-mandatory work started for 5
Canceling after 5
Canceling after 6
Canceling after 7
Canceling after 8
Non-mandatory work started for 9
Canceling after 9
```
</details>

# The problem
Still, there is a concurrency. We can observe it by making `non_mandatory_work` faster by removing `sleep` inside. Then, sometimes, `non_mandatory_work` finishes faster than `guard1.cancelled()` 

<details>

<summary>See output: </summary>

```
Non-mandatory work started for 0
Non-mandatory work done for 0
Mandatory work started for 0
^CMandatory work done for 0
Non-mandatory work started for 1
Non-mandatory work done for 1
Mandatory work started for 1
Mandatory work done for 1
Canceling after 2
Canceling after 3
Non-mandatory work started for 4
Non-mandatory work done for 4
Mandatory work started for 4
Mandatory work done for 4
Canceling after 5
Canceling after 6
Canceling after 7
Canceling after 8
Canceling after 9
```
</details>

# Naive solution
I solved this problem externally by creating an additional controller with Arc<AtomicBool> and infinite await. 

<details>
<summary> Naive implementation </summary>

```rust
use std::sync::Arc;
use std::sync::atomic::AtomicBool;
use std::time::Duration;
use tokio::signal::unix::signal;
use tokio::task;
use tokio::time::sleep;
use tokio_graceful::Shutdown;

#[tokio::main]
async fn main() {
    let shutdown = Shutdown::default();
    let guard1 = shutdown.guard();
    let controller = Arc::new(AtomicBool::new(false));
    let some = Some {
        controller: controller.clone(),
    };

    task::spawn(async move {
        for i in 0..10 {
            tokio::select! {
                _ = guard1.cancelled() => {
                    canceling(i).await;
                },
                result = some.non_mandatory_work(i) => {
                    mandatory_work(result).await;
                },
            }
        }
    });

    waiting_for_signal().await;
    controller.store(true, std::sync::atomic::Ordering::Relaxed);
    shutdown.shutdown().await;
}


struct Some {
    controller: Arc<AtomicBool>,
}

impl Some {
    async fn controller_fn(&self) {
        while self.controller.load(std::sync::atomic::Ordering::Relaxed) == true {
            sleep(Duration::from_secs(1)).await;
        }
    }

    async fn non_mandatory_work(&self, i: u32) -> u32 {
        self.controller_fn().await;
        println!("Non-mandatory work started for {}", i);
        println!("Non-mandatory work done for {}", i);
        i
    }

    fn shutdown(&self) {
        self.controller.store(true, std::sync::atomic::Ordering::Relaxed);
    }
}

async fn canceling(i: u32) {
    println!("Canceling after {}", i);
}


async fn mandatory_work(i: u32) {
    println!("Mandatory work started for {}", i);
    sleep(Duration::from_secs(2)).await;
    println!("Mandatory work done for {}", i);
}

async fn waiting_for_signal() {
    let mut sigterm = signal(tokio::signal::unix::SignalKind::terminate()).unwrap();
    let mut sighup = signal(tokio::signal::unix::SignalKind::hangup()).unwrap();
    let mut sigint = signal(tokio::signal::unix::SignalKind::interrupt()).unwrap();
    let mut sigquit = signal(tokio::signal::unix::SignalKind::quit()).unwrap();

    tokio::select! {
        _ = sigterm.recv() => {
            println!("Received SIGTERM");
        }
        _ = sighup.recv() => {
            println!("Received SIGHUP");
        }
        _ = sigint.recv() => {
            println!("Received SIGINT");
        }
        _ = sigquit.recv() => {
            println!("Received SIGQUIT");
        }
    }
}
```
</details>

But it will be nice to have infinite await in tokio-graceful crate. 

What do you think? Is this make sense to you? 

Thanks

**Comment by GlenDC** [comment]
I am not sure I  understand your issue due to the abstract example you gave here. But as far as I understand it you want to be able to:

- do concurrent work
- have the non-mandetory work stop ASAP when the cancellation has been kicked off
- but ensure that the mandetory work does conclude, with the results of the cancelled or finished mandetory work

If so then this is how you would do that with tokio-graceful as is (btw also notice that it's easier to spawn tasks directly with the graceful API as it will make it that you do not need to move/clone guards yourself and ensures all your tasks get a guard):

```rust
use std::time::Duration;
use tokio::time::sleep;
use tokio_graceful::{Shutdown, WeakShutdownGuard};

#[tokio::main]
async fn main() {
    let shutdown = Shutdown::default();
    shutdown.spawn_task_fn(|guard| async move {
        for i in 0..10 {
            let nm_guard = guard.clone_weak();
            let result = non_mandatory_work(nm_guard, i).await;
            mandatory_work(result).await;
        }
    });

    // Wait for shutdown signal
    shutdown.shutdown().await;
}

async fn canceling(i: u32) {
    println!("Canceling after {}", i);
}

async fn non_mandatory_work(guard: WeakShutdownGuard, i: u32) -> u32 {
    println!("Non-mandatory work started for {}", i);
    tokio::select! {
        _ = guard.cancelled() => {
            canceling(i).await;
            println!("Non-mandatory work cancelled early for {}", i);
            return i;
        }
        _ = sleep(Duration::from_secs(2)) => (),
    };
    println!("Non-mandatory work done for {}", i);
    i
}

async fn mandatory_work(i: u32) {
    println!("Mandatory work started for {}", i);
    sleep(Duration::from_secs(2)).await;
    println!("Mandatory work done for {}", i);
}
```

Which works and you can see that after CTRL+C that the non-mandatory works immediately returns because it was cancelled but that all mandatory work does get executed. Output:

```
Running `target/debug/i7`
Non-mandatory work started for 0
Non-mandatory work done for 0
Mandatory work started for 0
Mandatory work done for 0
Non-mandatory work started for 1
Non-mandatory work done for 1
Mandatory work started for 1
^CMandatory work done for 1
Non-mandatory work started for 2
Canceling after 2
Non-mandatory work cancelled early for 2
Mandatory work started for 2
Mandatory work done for 2
Non-mandatory work started for 3
Canceling after 3
Non-mandatory work cancelled early for 3
Mandatory work started for 3
Mandatory work done for 3
Non-mandatory work started for 4
Canceling after 4
Non-mandatory work cancelled early for 4
Mandatory work started for 4
Mandatory work done for 4
Non-mandatory work started for 5
Canceling after 5
Non-mandatory work cancelled early for 5
Mandatory work started for 5
Mandatory work done for 5
Non-mandatory work started for 6
Canceling after 6
Non-mandatory work cancelled early for 6
Mandatory work started for 6
Mandatory work done for 6
Non-mandatory work started for 7
Canceling after 7
Non-mandatory work cancelled early for 7
Mandatory work started for 7
Mandatory work done for 7
Non-mandatory work started for 8
Canceling after 8
Non-mandatory work cancelled early for 8
Mandatory work started for 8
Mandatory work done for 8
Non-mandatory work started for 9
Canceling after 9
Non-mandatory work cancelled early for 9
Mandatory work started for 9
Mandatory work done for 9
````

Final note: given that this does mean your cancellation gets delayed for until your non-mandatory is early cancelled and mandatory work is finished, that you :

- either have to wait infinite time for cancellation to complete (what you are currently doing)
- or cancel with timeout but put the timeout high enough (what I would recommend.

The second is achieved by using `shutdown_with_limit(Duration::from_secs(60))` instead of `shutdown.shutdown().await`. Higher time if you need more. Dunno how slow that graceful execution of your tasks is.

tokio-graceful was created so that you do not need to fiddle with signals and atomics yourself for this purpose. So hopefully my example here shows that you can easily achieve what you want, as-is, today already? Unless I misunderstood something?

**Comment by ctrlok** [comment]
Thank you so much for such a detailed answer. I'll try to explain what I'm trying to achieve because I feel that I previously explained it poorly. I'll use other terminology for the same steps to make them clearer. 

I have two scenarios: 

## Regular workflow
get a message from batch → process → submit 

## Regular workflow after shutdown
get a message from batch → send it back, don't start `process` and don't start `submit`

And I have two additional scenarios: 
if shutdown happens in the middle of `process` we send message back and don't run `submit` and don't wait `process` to finish

if shutdown happens in the middle of `submit` we continue with submitting

And overall get→process→submit and get batch→process→submit is typical scenarios 
<details>

```mermaid
sequenceDiagram
    participant RabbitMQ
    participant App
    participant Processor
    participant Submitter

    Note over RabbitMQ,Submitter: Regular workflow Start
    App->>RabbitMQ: Get batch of messages
    RabbitMQ-->>App: Batch of messages
    loop For each message in batch
        App->>Processor: Process message
        Processor-->>App: Processed message
        App->>Submitter: Submit processed message
        Submitter-->>App: Submission result
        App->>App: Mark message as done
    end
    App->>RabbitMQ: Get new batch of messages

    Note over RabbitMQ,Submitter: Regular Workflow End

    Note over RabbitMQ,Submitter: After Shutdown Signal Happens
    RabbitMQ-->>App: Batch of messages


    loop For each message in batch
        App->>App: Mark message as nack
        App->>RabbitMQ: Send message back
    end
    Note over RabbitMQ,Submitter: Shutdown Complete

    Note over RabbitMQ,Submitter: Additional Scenarios
    Note over App,Processor: Shutdown during process
    App->>Processor: Process message
    Processor-->>App: Shutdown received
    App->>App: Mark message as nack
    App->>RabbitMQ: Send message back

    Note over App,Submitter: Shutdown during submit
    App->>Submitter: Submit message
    Submitter-->>App: Shutdown received
    Submitter-->>App: Complete submission
    App->>App: Mark message as done
    App->>App: Proceed to next message
   ```
</details>

About code you posted: 

```
async fn non_mandatory_work(guard: WeakShutdownGuard, i: u32) -> u32 {
    println!("Non-mandatory work started for {}", i);
    tokio::select! {
        _ = guard.cancelled() => {
            canceling(i).await;
            println!("Non-mandatory work cancelled early for {}", i);
            return i;
        }
        _ = sleep(Duration::from_secs(2)) => (),
    };
    println!("Non-mandatory work done for {}", i);
    i
}
```

Here, `tokio::select!` has two branches: `guard.cancelled()` and `sleep(Duration::from_secs(2))` — in this demo it will work as expected, because `sleep(Duration::from_secs(2))` is a slow function and it returns after `guard.cancelled()` 

if we replace `sleep(Duration::from_secs(2))` to something like `async { println!("Hi") } ` — it will be fast enough to be selected before `guard.cancelled()` which leads to doing processing and submitting after shutdown signal was sent. 

And the only solution I can see right now is a function which do:

* Infinite await if we shuted down 
* pass if we not shuted down 

> either have to wait infinite time for cancellation to complete 

I'm not sure that I understand you there. Cancelation happens after branch with `guard.cancelled() ` is selected and that means that  other branch are dropped

**Comment by GlenDC** [comment]
About the last part I meant that on the actual `shutdown` await statement itself (so the only one in your app (section) that finishes on all work being completed). The `shutdown()` method waits forever, usually I prefer at least an upper-bound limit, before the orchestrating systems destroys it anyway.

So ok.. I think I understand you now... You mean that work might still get "started" as the `select` statement is fair and thus starts both and even might progress the 2nd without first checking if it was cancelled or not.

The simple solution is to put a select statement before with a timeout so that you are okay with it not being cancelled but just to check...

So with that approach you would get the following output:

```
Running `target/debug/i7`
Non-mandatory work started for 0
Silly work...
Non-mandatory work done for 0
Mandatory work started for 0
Mandatory work done for 0
Non-mandatory work started for 1
Silly work...
Non-mandatory work done for 1
Mandatory work started for 1
^CMandatory work done for 1
Canceling after 2
Non-mandatory work not started because of guard cancellation for 2
Mandatory work started for 2
Mandatory work done for 2
Canceling after 3
Non-mandatory work not started because of guard cancellation for 3
Mandatory work started for 3
Mandatory work done for 3
Canceling after 4
Non-mandatory work not started because of guard cancellation for 4
Mandatory work started for 4
Mandatory work done for 4
Canceling after 5
Non-mandatory work not started because of guard cancellation for 5
Mandatory work started for 5
Mandatory work done for 5
Canceling after 6
Non-mandatory work not started because of guard cancellation for 6
Mandatory work started for 6
Mandatory work done for 6
Canceling after 7
Non-mandatory work not started because of guard cancellation for 7
Mandatory work started for 7
Mandatory work done for 7
Canceling after 8
Non-mandatory work not started because of guard cancellation for 8
Mandatory work started for 8
Mandatory work done for 8
Canceling after 9
Non-mandatory work not started because of guard cancellation for 9
Mandatory work started for 9
Mandatory work done for 9
```

with the updated code:

```rust
use std::time::Duration;
use tokio::time::sleep;
use tokio_graceful::{Shutdown, WeakShutdownGuard};

#[tokio::main]
async fn main() {
    let shutdown = Shutdown::default();
    shutdown.spawn_task_fn(|guard| async move {
        for i in 0..10 {
            let nm_guard = guard.clone_weak();
            let result = non_mandatory_work(nm_guard, i).await;
            mandatory_work(result).await;
        }
    });

    // Wait for shutdown signal
    shutdown.shutdown().await;
}

async fn canceling(i: u32) {
    println!("Canceling after {}", i);
}

async fn non_mandatory_work(guard: WeakShutdownGuard, i: u32) -> u32 {
    tokio::select! {
        _ = guard.cancelled() => {
            canceling(i).await;
            println!("Non-mandatory work not started because of guard cancellation for {}", i);
            return i;
        }
        _timeout = sleep(Duration::from_micros(1)) => (),
    };

    println!("Non-mandatory work started for {}", i);
    let silly_fut = async {
        println!("Silly work...");
    };
    tokio::select! {
        _ = guard.cancelled() => {
            canceling(i).await;
            println!("Non-mandatory work cancelled early for {}", i);
            return i;
        }
        _ = silly_fut => (),
    };
    println!("Non-mandatory work done for {}", i);
    i
}

async fn mandatory_work(i: u32) {
    println!("Mandatory work started for {}", i);
    sleep(Duration::from_secs(2)).await;
    println!("Mandatory work done for {}", i);
}
```

**Comment by GlenDC** [comment]
That said, while for many use cases that silly wait time is not an issue, it's also not really technically needed, more of a limitation of the API. So if I do understand you correctly _now_ and the above does achieve what you want then I can make a new version where

```rust
tokio::select! {
        _ = guard.cancelled() => {
            canceling(i).await;
            println!("Non-mandatory work not started because of guard cancellation for {}", i);
            return i;
        }
        _timeout = sleep(Duration::from_micros(1)) => (),
    };
```
can be replaced with:

```rust
if guard.cancelled_peek() {
        canceling(i).await;
        println!("Non-mandatory work not started because of guard cancellation for {}", i);
        return i;
    }
```

Which will ensure to get the same behaviour without having to also add an arbitrary delay in there.

Also if you want that you can perhaps help me come up with a different name then `cancelled_peek` as it sounds a bit strange...

**Comment by GlenDC** [comment]
Got a branch open for you that I can use as a starting point: https://github.com/plabayo/tokio-graceful/pull/8

it's however not yet working so would need some work, but I imagine there is a way, if that is indeed a kind of feature you would need. Output currently shows though that this approach does not work yet:

```
 Running `target/debug/i7`
Non-mandatory work started for 0
Silly work...
Non-mandatory work done for 0
Mandatory work started for 0
Mandatory work done for 0
Non-mandatory work started for 1
Silly work...
Non-mandatory work done for 1
Mandatory work started for 1
^CMandatory work done for 1
Non-mandatory work started for 2
Canceling after 2
Non-mandatory work cancelled early for 2
Mandatory work started for 2
Mandatory work done for 2
Non-mandatory work started for 3
Canceling after 3
Non-mandatory work cancelled early for 3
Mandatory work started for 3
Mandatory work done for 3
Non-mandatory work started for 4
Silly work...
Non-mandatory work done for 4
Mandatory work started for 4
Mandatory work done for 4
Non-mandatory work started for 5
Silly work...
Non-mandatory work done for 5
Mandatory work started for 5
Mandatory work done for 5
Non-mandatory work started for 6
Silly work...
Non-mandatory work done for 6
Mandatory work started for 6
Mandatory work done for 6
Non-mandatory work started for 7
Silly work...
Non-mandatory work done for 7
Mandatory work started for 7
Mandatory work done for 7
Non-mandatory work started for 8
Silly work...
Non-mandatory work done for 8
Mandatory work started for 8
Mandatory work done for 8
Non-mandatory work started for 9
Canceling after 9
Non-mandatory work cancelled early for 9
Mandatory work started for 9
Mandatory work done for 9
```

**Comment by GlenDC** [comment]
ok @ctrlok added some more code seems to work now in the example code above:

```
Running `target/debug/i7`
Non-mandatory work started for 0
Silly work...
Non-mandatory work done for 0
Mandatory work started for 0
Mandatory work done for 0
Non-mandatory work started for 1
Silly work...
Non-mandatory work done for 1
Mandatory work started for 1
^CMandatory work done for 1
Canceling after 2
Non-mandatory work not started because of guard cancellation for 2
Mandatory work started for 2
Mandatory work done for 2
Canceling after 3
Non-mandatory work not started because of guard cancellation for 3
Mandatory work started for 3
Mandatory work done for 3
Canceling after 4
Non-mandatory work not started because of guard cancellation for 4
Mandatory work started for 4
Mandatory work done for 4
Canceling after 5
Non-mandatory work not started because of guard cancellation for 5
Mandatory work started for 5
Mandatory work done for 5
Canceling after 6
Non-mandatory work not started because of guard cancellation for 6
Mandatory work started for 6
Mandatory work done for 6
Canceling after 7
Non-mandatory work not started because of guard cancellation for 7
Mandatory work started for 7
Mandatory work done for 7
Canceling after 8
Non-mandatory work not started because of guard cancellation for 8
Mandatory work started for 8
Mandatory work done for 8
Canceling after 9
Non-mandatory work not started because of guard cancellation for 9
Mandatory work started for 9
Mandatory work done for 9
```

Is that something that would "unblock" you without needing to resort to workarounds. Or does this not help you? if it doesn't please elaborate further on my example (E.g. edit it further to show where there's still an issue for you).

If this does solve your issue then I'll add some more Q&A and testing code and prepare a release today.

**Comment by ctrlok** [comment]
Hey! 

Thanks for the answers. I will try your implementation, but I found out that it's unnecessary. 

tokio::select! macro [supports](https://docs.rs/tokio/latest/tokio/macro.select.html#fairness) `biased;` option. 

Normaly tokio::select! runs branches randomly. But with the `biased;` option, it runs branches from top to bottom, and the shutdown check always runs first. The `non-mandatory` never has a chance to run if the shutdown has already happened. And it perfectly fits my case.

**Comment by ctrlok** [comment]
I checked the cancelled_peek() functionality, and it's pretty interesting and useful. Still, it didn't fit my case because it requires borrowing guard as mutable while I'm trying to borrow it as immutable at the same time (for `guard1.cancelled()` in the first branch) 

It could be solved by adding a weak guard, but in my tests, it didn't work. 

For code: 

```rust
    let shutdown = Shutdown::default();
    let guard1 = shutdown.guard();
    let mut week = shutdown.guard_weak();

    task::spawn(async move {
        for i in 0..10 {
            tokio::select! {
                _ = guard1.cancelled() => {
                    canceling(i).await;
                },
                result = async {
                    if weak.cancelled_peek() {
                        println!("Weak cancelled for {}", i);
                        loop {}
                    }
                    non_mandatory_work(i).await
                } => {
                    mandatory_work(result).await;
                },
            }
        }
    });
```

it produces output 
```
Non-mandatory work started for 0
^CReceived SIGINT
Canceling after 0
Canceling after 1
Weak cancelled for 2
```
and goes to the infinite loop after it 

That means it somehow blocks `guard1.cancelled()` from execution, and I don't understand why because same `loop {}` works in case I'm using additional `Arc<AtomicBool>` to store shutdown state

**Comment by GlenDC** [comment]
> https://docs.rs/tokio/latest/tokio/macro.select.html#fairness

Perfect @ctrlok then I think I'm going to close the open MR and drop that feature. As it's added complexity/code, without a real use case then. We can always add it later if we do find a use case for which it is required. Seems to me that that biased does indeed achieve what you wanted to do.

**Comment by ctrlok** [comment]
Thank you so much for such detailed answers and help!

**Comment by GlenDC** [comment]
> Still, it didn't fit my case because it requires borrowing guard as mutable while I'm trying to borrow it as immutable at the same time (for guard1.cancelled() in the first branch)

I do not get what the issue with requiring mutability though. `tokio-graceful` is created to make cloning guards cheap (e.g. non-locking). So feel free to make as many clones as you want. That goes the same for regular guards as well as weak guards. So as such you should be able to mark it as `mut` given that you can have always a single owner.

That said, if we do ever find a use case for the peeking, we can also make it `&self` instead of `&mut self`, as the core functionality for peek doesn't require mutability. I only did it so that I can update the state if I checked it anyway, but that's more of an optimisation, not a requirement. So to be clear, we can make it immutable if also that has a real use case.

And as I said before in my examples do make your life easier by spawning tasks or task functions directly from the guard or main shutdown handler. It does the same as `task::spawn` but it ensures there is a guard cloned for you and moved into the task without you having to do that manual work.

The infinite loop is normal though, as you never await within the loops, so you are essentially doing a blocking task within your future. Which is a mix of colours. And not a good one. First of all that's a very expensive operation either way, as it's similar to a "spinning lock", but if you do for some reason I do not get want something like that you'll at the very least will need to `yield` (https://docs.rs/tokio/latest/tokio/task/fn.yield_now.html) on every iteration, giving other futures the chance to poll. Even then though that future will never complete making me even less sure why you want something like that.

Glad to help. Hopefully you are indeed helped and unblocked, and can happily use tokio-graceful onwards and forward. If not, do feel free to re-open or make a new issue. I do want to help, and gladly do so. So if you feel stuck or have a question do feel free to reach out :) Take care.
